Realm/ObjectServerTests/SwiftObjectServerTests.swift

```diff
@@ -3079,6 +2223,124 @@ class AsyncAwaitObjectServerTests: SwiftSyncTestCase {
+        try await client.deregisterDevice(user: app.currentUser!)
+        XCTAssertTrue(true)
+    }
+
+    func testEmailPasswordProviderClientAsyncAwait() async throws {
+        let email = "realm_tests_do_autoverify\(randomString(7))@\(randomString(7)).com"
+        let password = randomString(10)
+        try await app.emailPasswordAuth.registerUser(email: email, password: password)
+
+        do {
+            try await app.emailPasswordAuth.confirmUser("atoken", tokenId: "atokenid")
+        } catch {
+            XCTAssertNotNil(error)
+        }
+
+        do {
+            try await app.emailPasswordAuth.resendConfirmationEmail(email)
+        } catch {
+            XCTAssertNotNil(error)
+        }
+
+        do {
+            try await app.emailPasswordAuth.retryCustomConfirmation(email)
+        } catch {
+            XCTAssertNotNil(error)
+        }
+
+        do {
+            try await app.emailPasswordAuth.sendResetPasswordEmail("atoken")
+        } catch {
+            XCTAssertNotNil(error)
+        }
+    }
+
+    func testUserAPIKeyProviderClientAsyncAwait() async throws {
+        let email = "realm_tests_do_autoverify\(randomString(7))@\(randomString(7)).com"
+        let password = randomString(10)
+        try await app.emailPasswordAuth.registerUser(email: email, password: password)
+
+        let credentials = Credentials.emailPassword(email: email, password: password)
+        let syncUser = try await self.app.login(credentials: credentials)
+        let apiKey = try await syncUser.apiKeysAuth.createAPIKey(named: "my-api-key")
+        XCTAssertNotNil(apiKey)
+
+        let fetchedApiKey = try await syncUser.apiKeysAuth.fetchAPIKey(apiKey.objectId)
+        XCTAssertNotNil(fetchedApiKey)
+
+        let fetchedApiKeys = try await syncUser.apiKeysAuth.fetchAPIKeys()
+        XCTAssertNotNil(fetchedApiKeys)
+        XCTAssertEqual(fetchedApiKeys.count, 1)
+
+        try await syncUser.apiKeysAuth.disableAPIKey(apiKey.objectId)
+        try await syncUser.apiKeysAuth.enableAPIKey(apiKey.objectId)
+        try await syncUser.apiKeysAuth.deleteAPIKey(apiKey.objectId)
+
+        let newFetchedApiKeys = try await syncUser.apiKeysAuth.fetchAPIKeys()
+        XCTAssertNotNil(newFetchedApiKeys)
+        XCTAssertEqual(newFetchedApiKeys.count, 0)
+    }
+
+    func testCustomUserDataAsyncAwait() async throws {
+        let email = "realm_tests_do_autoverify\(randomString(7))@\(randomString(7)).com"
+        let password = randomString(10)
+        try await app.emailPasswordAuth.registerUser(email: email, password: password)
+
+        let syncUser = try await self.app.login(credentials: Credentials.anonymous)
+        XCTAssertNotNil(syncUser)
+
+        let userDataEx = expectation(description: "Update user data")
+        app.currentUser?.functions.updateUserData([["favourite_colour": "green", "apples": 10]]) { _, error  in
+            XCTAssertNil(error)
+            userDataEx.fulfill()
+        }
```

Comment:
Shouldn't this have an async version too?

Reply:
looks like subscripts cannot be `async` as mentioned in the proposal https://github.com/apple/swift-evolution/blob/main/proposals/0296-async-await.md#:~:text=for%20such%20references.-,Special,-functions%20like%20deinit, maybe they will allow it in the future for subscripts that only have a getter which is the case here

Reply:
There is a clever, non-breaking way to make this `async` that needs to be accomplished in this PR. Since our Functions struct only returned pseudo completionHandler continuations, we can overload another structure that leverages `dynamicCallable` that can then be async:

```swift
// end result
func testCallFunctionAsync() async throws {
    let user = try await app.login(credentials: basicCredentials())

    guard case let .int32(sum) = try await user.functions.sum([1, 2, 3, 4, 5]) else {
        return XCTFail("Should be int32")
    }
    XCTAssertEqual(sum, 15)

    guard case let .bool(upd) = try await user.functions.updateUserData([
        ["favourite_colour": "green", "apples": 10]
    ]) else {
        return XCTFail("Should be bool")
    }
    XCTAssertTrue(upd)
}


// implementation
@dynamicMemberLookup
@frozen public struct Functions {

    private let user: User

    fileprivate init(user: User) {
        self.user = user
    }

    /// A closure type for receiving the completion of a remote function call.
    public typealias FunctionCompletionHandler = (AnyBSON?, Error?) -> Void

    /// A closure type for the dynamic remote function type.
    public typealias Function = ([AnyBSON], @escaping FunctionCompletionHandler) -> Void

    /// The implementation of @dynamicMemberLookup that allows for dynamic remote function calls.
    public subscript(dynamicMember string: String) -> Function {
        return { (arguments: [AnyBSON], completionHandler: @escaping FunctionCompletionHandler) in
            let objcArgs = arguments.map(ObjectiveCSupport.convertBson)
            self.user.__callFunctionNamed(string, arguments: objcArgs) { (bson: RLMBSON?, error: Error?) in
                completionHandler(bson.map(ObjectiveCSupport.convertBson) ?? .none, error)
            }
        }
    }

    /// The implementation of @dynamicMemberLookup that allows for dynamic remote function calls.
    public subscript(dynamicMember string: String) -> FunctionCallable {
        FunctionCallable(name: string, user: user)
    }
}

@dynamicCallable
public struct FunctionCallable {
    fileprivate let name: String
    fileprivate let user: User

    public func dynamicallyCall(withArguments args: [([AnyBSON], ((Result<AnyBSON, Error>) -> Void))]) -> Void {
        let objcArgs = args.first!.0.map(ObjectiveCSupport.convertBson)
        self.user.__callFunctionNamed(name, arguments: objcArgs) { (bson: RLMBSON?, error: Error?) in
            if let b = bson.map(ObjectiveCSupport.convertBson), let bson = b {
                args.first!.1(.success(bson))
            } else {
                args.first!.1(.failure(error ?? Realm.Error.callFailed))
            }
        }
    }

    #if !(os(iOS) && (arch(i386) || arch(arm)))
    /// The implementation of @dynamicMemberLookup that allows for dynamic remote function calls.
    @available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, macCatalyst 13.0, macCatalystApplicationExtension 13.0, *)
    public func dynamicallyCall(withArguments args: [[AnyBSON]]) -> Future<AnyBSON, Error> {
        return Future<AnyBSON, Error> { promise in
            let objcArgs = args.first!.map(ObjectiveCSupport.convertBson)
            self.user.__callFunctionNamed(name, arguments: objcArgs) { (bson: RLMBSON?, error: Error?) in
                if let b = bson.map(ObjectiveCSupport.convertBson), let bson = b {
                    promise(.success(bson))
                } else {
                    promise(.failure(error ?? Realm.Error.callFailed))
                }
            }
        }
    }
    #endif
}

@available(macOS 12.0, tvOS 15.0, iOS 15.0, watchOS 8.0, *)
extension FunctionCallable {
    public func dynamicallyCall(withArguments args: [[AnyBSON]]) async throws -> AnyBSON {
        try await withCheckedThrowingContinuation { continuation in
            let objcArgs = args.first!.map(ObjectiveCSupport.convertBson)
            self.user.__callFunctionNamed(name, arguments: objcArgs) { (bson: RLMBSON?, error: Error?) in
                if let b = bson.map(ObjectiveCSupport.convertBson), let bson = b {
                    continuation.resume(returning: bson)
                } else {
                    continuation.resume(throwing: error ?? Realm.Error.callFailed)
                }
            }
        }
    }
}
```

Note that this code has some unnecessary repetition that could use cleaning up, but the rest stands.